tools/istio-iptables/pkg/dependencies/implementation.go

```diff
@@ -51,51 +50,133 @@ var exittypeToString = map[XTablesExittype]string{
+	DetectedSaveBinary    string
+	DetectedRestoreBinary string
 	// the actual version
-	version *utilversion.Version
+	Version *utilversion.Version
 	// true if legacy mode, false if nf_tables
-	legacy bool
+	Legacy bool
+	// true if we detected that existing rules are present for this variant (legacy, nft, v6)
+	ExistingRules bool
 }
 
-func DetectIptablesVersion(ver string) (IptablesVersion, error) {
-	if ver == "" {
-		var err error
-		verb, err := exec.Command("iptables", "--version").CombinedOutput()
-		if err != nil {
-			return IptablesVersion{}, err
-		}
-		ver = string(verb)
+func (v IptablesVersion) CmdToString(cmd constants.IptablesCmd) string {
+	switch cmd {
+	case constants.IPTables:
+		return v.DetectedBinary
+	case constants.IPTablesSave:
+		return v.DetectedSaveBinary
+	case constants.IPTablesRestore:
+		return v.DetectedRestoreBinary
+	default:
+		return ""
+	}
+}
+
+// IsWriteCmd returns true for all command types that do write actions (and thus need a lock)
+func (v IptablesVersion) IsWriteCmd(cmd constants.IptablesCmd) bool {
+	switch cmd {
+	case constants.IPTables:
+		return true
+	case constants.IPTablesRestore:
+		return true
+	default:
+		return false
+	}
+}
+
+// Constants for iptables commands
+// These should not be used directly/assumed to be present, but should be contextually detected
+const (
+	iptablesBin         = "iptables"
+	iptablesNftBin      = "iptables-nft"
+	iptablesLegacyBin   = "iptables-legacy"
+	ip6tablesBin        = "ip6tables"
+	ip6tablesNftBin     = "ip6tables-nft"
+	ip6tablesLegacyBin  = "ip6tables-legacy"
+	iptablesRestoreBin  = "iptables-restore"
+	ip6tablesRestoreBin = "ip6tables-restore"
+)
+
+// It is not sufficient to check for the presence of one binary or the other in $PATH -
+// we must choose a binary that is
+// 1. Available in our $PATH
+// 2. Matches where rules are actually defined in the netns we're operating in
+// (legacy or nft, with a preference for the latter if both present)
+//
+// This is designed to handle situations where, for instance, the host has nft-defined rules, and our default container
+// binary is `legacy`, or vice-versa - we must match the binaries we have in our $PATH to what rules are actually defined
+// in our current netns context.
+//
+// Q: Why not simply "use the host default binary" at $PATH/iptables?
+// A: Because we are running in our own container and do not have access to the host default binary.
+// We are using our local binaries to update host rules, and we must pick the right match.
+//
+// Basic selection logic is as follows:
+// 1. see if we have `nft` binary set in our $PATH
+// 2. see if we have existing rules in `nft` in our netns
+// 3. If so, use `nft` binary set
+// 4. Otherwise, see if we have `legacy` binary set, and use that.
+// 5. Otherwise, see if we have `iptables` binary set, and use that (detecting whether it's nft or legacy).
+func (r *RealDependencies) DetectIptablesVersion(ipV6 bool) (IptablesVersion, error) {
```

Comment:
should we have some logging around this? (I may have missed it)

Reply:
Only thing that stopped me was that the logger wasn't already imported here, but I might just add it and put in some logging, yeah.